sync/syncer.go

```diff
@@ -291,149 +411,133 @@ func (s *Syncer) Txs(mb *types.Block) ([]*types.SerializableTransaction, error)
+			return nil, errors.New(fmt.Sprintf("could not fetch tx %v", hex.EncodeToString(id[:])))
 		}
 	}
 	return txs, nil
 }
 
-func (s *Syncer) ATXs(mb *types.Block) (atxs []*types.ActivationTx, associated *types.ActivationTx, err error) {
+func (s *Syncer) syncAtxs(atxIds []types.AtxId) (atxs []*types.ActivationTx, err error) {
 
 	//look in pool
-	atxMap := make(map[types.AtxId]*types.ActivationTx, len(atxs))
+	poolAtxs := make(map[types.AtxId]*types.ActivationTx, len(atxs))
 	missingInPool := make([]types.AtxId, 0, len(atxs))
-	for _, t := range mb.AtxIds {
+	for _, t := range atxIds {
 		id := t
 		if x := s.atxpool.Get(id); x != nil {
 			atx := x.(*types.ActivationTx)
 			if atx.Nipst == nil {
-				s.Warning("atx %v nipst not found ", hex.EncodeToString(id.Bytes()))
+				s.Warning("atx %v nipst not found ", id.ShortId())
 				missingInPool = append(missingInPool, id)
 				continue
 			}
-			s.Info("found atx, %v in atx pool", hex.EncodeToString(id.Bytes()))
-			atxMap[id] = atx
+			s.Info("found atx, %v in atx pool", id.ShortId())
+			poolAtxs[id] = atx
 		} else {
-			s.Warning("atx %v not in atx pool", hex.EncodeToString(id.Bytes()))
+			s.Warning("atx %v not in atx pool", id.ShortId())
 			missingInPool = append(missingInPool, id)
 		}
 	}
 
 	//look in db
-	dbAtxs, missing := s.GetATXs(missingInPool)
-
-	fetchAssociated := false
-	if mb.ATXID != *types.EmptyAtxId {
-		if _, err := s.GetAtx(mb.ATXID); err != nil {
-			fetchAssociated = true
-			missing = append(missing, mb.ATXID)
-		}
-	}
+	dbAtxs, missingInDb := s.GetATXs(missingInPool)
 
 	//map and sort atxs
-	if len(missing) > 0 {
-		output := s.fetchWithFactory(ATxReqFactory(missing), 1)
+	if len(missingInDb) > 0 {
+
+		output := s.fetchWithFactory(NewNeighborhoodWorker(s, 1, ATxReqFactory(missingInDb)))
 		for out := range output {
-			ntxs := out.([]types.ActivationTx)
-			for _, atx := range ntxs {
-				atxMap[atx.Id()] = &atx
+			atxs := out.([]types.ActivationTx)
+			for _, atx := range atxs {
+				if err := s.ValidateAtx(&atx); err != nil {
+					s.Warning("atx %v not valid %v", atx.ShortId(), err)
+					continue
+				}
+				poolAtxs[atx.Id()] = &atx
 			}
 		}
 	}
 
-	if fetchAssociated {
-		if _, ok := atxMap[mb.BlockHeader.ATXID]; !ok {
-			return nil, nil, errors.New(fmt.Sprintf("could not fetch associated %v", hex.EncodeToString(mb.BlockHeader.ATXID.Bytes())))
-		}
-	}
-
-	atxs = make([]*types.ActivationTx, 0, len(mb.AtxIds))
-	for _, t := range mb.AtxIds {
-		if tx, ok := atxMap[t]; ok {
+	atxs = make([]*types.ActivationTx, 0, len(atxIds))
+	for _, id := range atxIds {
+		if tx, ok := poolAtxs[id]; ok {
 			atxs = append(atxs, tx)
-		} else if tx, ok := dbAtxs[t]; ok {
+		} else if tx, ok := dbAtxs[id]; ok {
 			atxs = append(atxs, tx)
 		} else {
-			return nil, nil, errors.New(fmt.Sprintf("could not fetch atx %v", hex.EncodeToString(t.Bytes())))
+			return nil, errors.New(fmt.Sprintf("could not fetch atx %v", id.ShortId()))
```

Comment:
IMO you should try to fetch all of them and then print which ATX's were missing, it will make debugging easier